### PR TITLE
Build MANA on Perlmutter

### PR DESCRIPTION
**Pull request by dahongli**
This includes several changes,
[1] Port Twinkle's changes of supporting MANA on Perlmutter to main branch
[2] Use static libraries on "cray" version Linux release. This allows us to use the different build options on Perlmutter and Cori so we can build MANA on both platform.
[3] Choose the right MPI_Type API based on MPICH version.

### CODE REVIEW

File: mpi-proxy-split/lower-half/static_libs.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+ -lmpi_gnu_91 -lpmi /opt/cray/libfabric/1.15.0.0/lib64/libfabric.a -lcxi -ldl -latomic -lrt -lpthread /opt/cray/pe/pals/1.1.8/lib/libpals.a /global/homes/z/zz217/sw/opt/spack/cray-sles15-zen3/curl-7.80.0-vxd53vd/lib/libcurl.a /global/homes/z/zz217/sw/opt/spack/cray-sles15-zen3/readline-8.1-ib3nkkj/lib/libreadline.a /global/homes/z/zz217/sw/opt/spack/cray-sles15-zen3/readline-8.1-ib3nkkj/lib/libhistory.a /global/homes/z/zz217/sw/opt/spack/cray-sles15-zen3/zlib-1.2.11-nh5csvy/lib/libz.a /global/homes/z/zz217/sw/opt/spack/cray-sles15-zen3/openssl-1.1.1m-5yvmvee/lib/libssl.a /global/homes/z/zz217/sw/opt/spack/cray-sles15-zen3/openssl-1.1.1m-5yvmvee/lib/libcrypto.a /global/homes/z/zz217/sw/opt/spack/cray-sles15-zen3/bzip2-1.0.8-4p4fezv/lib/libbz2.a /global/homes/z/zz217/sw/opt/spack/cray-sles15-zen3/json-c-0.15-avwb2j6/lib64/libjson-c.a /global/homes/z/zz217/sw/opt/spack/cray-sles15-zen3/libzip-1.2.0-rolalf6/lib/libzip.a /global/homes/z/zz217/sw/opt/spack/cray-sles15-zen3/gdbm-1.19-v5pbqq3/lib/libgdbm_compat.a /global/homes/z/zz217/sw/opt/spack/cray-sles15-zen3/gdbm-1.19-v5pbqq3/lib/libgdbm.a -Wl,--end-group
```
Comment (gc00) [inline mpi-proxy-split/lower-half/static_libs.txt]: Please remov `-Wl,--end-group` at the end.  See the comment about the Makefile for why.

File: mpi-proxy-split/mpi-wrappers/generate-mpi-fortran-wrappers.py [hunk 2]
```diff
@@ -25,7 +25,7 @@
 declarations_file.close()
 
 def print_mpi_type_struct():
-  print("#ifdef CRAY_MPICH_VERSION")
+  print("#if MPICH_NUMVERSION >= MPICH_CALC_VERSION(3,4,0,0,2) && defined(CRAY_MPICH_VERSION)")
```
Comment (gc00) [inline mpi-proxy-split/mpi-wrappers/generate-mpi-fortran-wrappers.py]: Why MPICH 3.4?  Can we add a short comment about what changed in that version.

File: mpi-proxy-split/lower-half/long-static_libs.txt [hunk 3]
```diff
@@ -0,0 +1 @@
+ -lmpi_gnu_91 -lpmi /opt/cray/libfabric/1.15.0.0/lib64/libfabric.a -lcxi -ldl -latomic -lrt -lpthread /opt/cray/pe/pals/1.1.8/lib/libpals.a /global/homes/z/zz217/sw/opt/spack/cray-sles15-zen3/curl-7.80.0-vxd53vd/lib/libcurl.a /global/homes/z/zz217/sw/opt/spack/cray-sles15-zen3/readline-8.1-ib3nkkj/lib/libreadline.a /global/homes/z/zz217/sw/opt/spack/cray-sles15-zen3/readline-8.1-ib3nkkj/lib/libhistory.a /global/homes/z/zz217/sw/opt/spack/cray-sles15-zen3/zlib-1.2.11-nh5csvy/lib/libz.a /global/homes/z/zz217/sw/opt/spack/cray-sles15-zen3/openssl-1.1.1m-5yvmvee/lib/libssl.a /global/homes/z/zz217/sw/opt/spack/cray-sles15-zen3/openssl-1.1.1m-5yvmvee/lib/libcrypto.a /global/homes/z/zz217/sw/opt/spack/cray-sles15-zen3/bzip2-1.0.8-4p4fezv/lib/libbz2.a /global/homes/z/zz217/sw/opt/spack/cray-sles15-zen3/json-c-0.15-avwb2j6/lib64/libjson-c.a /global/homes/z/zz217/sw/opt/spack/cray-sles15-zen3/libzip-1.2.0-rolalf6/lib/libzip.a /global/homes/z/zz217/sw/opt/spack/cray-sles15-zen3/gdbm-1.19-v5pbqq3/lib/libgdbm_compat.a /global/homes/z/zz217/sw/opt/spack/cray-sles15-zen3/gdbm-1.19-v5pbqq3/lib/libgdbm.a -Wl,--end-group
```
Comment (gc00) [inline mpi-proxy-split/lower-half/long-static_libs.txt:1]: Ren=move `-Wl,--end-group` here and it back into any Makefile that uses `long-static-libs.txt`.

File: mpi-proxy-split/split_process.cpp [hunk 4]
```diff
@@ -261,10 +263,13 @@ read_lh_proxy_bits(pid_t childpid)
     ret = process_vm_readv(childpid, remote_iov + i, 1, remote_iov + i, 1, 0);
     JASSERT(ret != -1)(JASSERT_ERRNO).Text("Error reading data from lh_proxy");
     // Can remove PROT_WRITE now that we've populated the segment.
-    ret = mprotect(remote_iov[i].iov_base, remote_iov[i].iov_len,
-                  lh_regions_list[i].prot);
+    // Update: on perlmutter, we need to keep the PROT_WRITE permission as
```
Comment (gc00) [inline mpi-proxy-split/split_process.cpp:266]: perlmutter -> Perlmutter

File: mpi-proxy-split/split_process.cpp [hunk 5]
```diff
@@ -261,10 +263,13 @@ read_lh_proxy_bits(pid_t childpid)
...
+    // libc writes to a read-only region for some reason. Keeping PROT_WRITE is
+    // is harmless.
```
Comment (gc00) [inline mpi-proxy-split/split_process.cpp:268]: Remove `is`.  It already appears on line above.

File: mpi-proxy-split/lower-half/static_libs.txt [hunk 6]
```diff
@@ -1 +1 @@
--lmpi_gnu_91 -lpmi -lpmi2 /opt/cray/libfabric/1.15.0.0/lib64/libfabric.a -lcxi -L/global/homes/z/zz217/local/lib -ljson-c  -lcurl  -lssl -lcrypto -lz -latomic -lpthread /opt/cray/pe/pals/1.1.8/lib/libpals.a -Wl,--end-group
+-lpmi -lpmi2 /opt/cray/libfabric/1.15.0.0/lib64/libfabric.a -lcxi -L/global/homes/z/zz217/local/lib -ljson-c  -lcurl  -lssl -lcrypto -lz -latomic -lpthread /opt/cray/pe/pals/1.1.8/lib/libpals.a -Wl,--end-group
```
Comment (gc00) [inline mpi-proxy-split/lower-half/static_libs.txt]: Here also, please remove the `-Wl,--end-group` at the end, and add it back into any Makefile that uese `static_libs.txt` from lower-half subdirectory.

File: mpi-proxy-split/mpi-wrappers/generate-mpi-fortran-wrappers.py [hunk 7]
```diff
@@ -44,7 +44,7 @@ def print_mpi_type_struct():
   print("}")
 
 def print_mpi_type_hindexed():
-  print("#ifdef CRAY_MPICH_VERSION")
+  print("#if MPICH_NUMVERSION < MPICH_CALC_VERSION(3,4,0,0,2) && defined(CRAY_MPICH_VERSION)")
```
Comment (gc00) [inline mpi-proxy-split/mpi-wrappers/generate-mpi-fortran-wrappers.py:47]: Let's add a brief comment about what changed i MPICH-3.4.0.2

**Review by gc00 — CHANGES_REQUESTED** [review]
This is close, but could you make the changes requested?  After that\t I can quickly approve this.

File: mpi-proxy-split/mpi-wrappers/generate-mpi-fortran-wrappers.py (hunk 2 above)
Comment (dahongli) [reply mpi-proxy-split/mpi-wrappers/generate-mpi-fortran-wrappers.py]: @JainTwinkle Any idea why checking MPICH_CALC_VERSION(3,4,0,0,2)? I saw this checking was first added for api_type_struct for build mana on Perlmutter.

File: mpi-proxy-split/lower-half/static_libs.txt (hunk 1 above)
Comment (dahongli) [reply mpi-proxy-split/lower-half/static_libs.txt]: I fixed it. Thanks Gene!

File: mpi-proxy-split/mpi-wrappers/generate-mpi-fortran-wrappers.py (hunk 2 above)
Comment (dahongli) [reply mpi-proxy-split/mpi-wrappers/generate-mpi-fortran-wrappers.py]: MPICH 3.0 version has already used the same API. Why we choose version 3.4a2?

File: mpi-proxy-split/lower-half/static_libs.txt (hunk 6 above)
Comment (dahongli) [reply mpi-proxy-split/lower-half/static_libs.txt]: Done.

File: mpi-proxy-split/lower-half/long-static_libs.txt (hunk 3 above)
Comment (gc00) [reply mpi-proxy-split/lower-half/long-static_libs.txt:1]: @dahongli , Could you also remove `-Wl,--end-grou` from this long-static-libs.txt file?  The principle should be that we put `begin--grup` and `end-group` _only_ in Makefile's and not in files of type static-libs.txt.  Thanks.

**Review by gc00 — CHANGES_REQUESTED** [review]
The new README.md looks good.  Thanks.

And thanks for adjusting the `-Wll,-end-group` expression.  I thinnk there's one place where you missed modifying it:  long-static--libs.txt.  You'll see that in my inline comment.s

Aside from that let's wait to hear from Twinkle if there something special about MPICH-3.14, or if it's really about MPICH-30.  Assuming it's MPICH-30, you can make that one further change, and then I will quickly review and approve this PR.

Thanks.

File: mpi-proxy-split/mpi-wrappers/generate-mpi-fortran-wrappers.py (hunk 2 above)
Comment (JainTwinkle) [reply mpi-proxy-split/mpi-wrappers/generate-mpi-fortran-wrappers.py]: Hi @dahongli and @gc00,
I was getting a compile-time error on Perlmutter with the existing macro that checks for `CRAY_MPICH_VERSION`. This is required to differentiate between a cray and a non-cray platform. However, that was not sufficient for Perlmutter as it uses the else part signature (`(MPI_Datatype*) array_of_types` argument without the `const`). So, at that time, I needed to differentiate between Cori's cray-mpich and Perlmutter's cray-mpich, to ensure that the else section API is being used when running on Perlmutter.
Feel free to improve the if...else condition.

File: mpi-proxy-split/mpi-wrappers/generate-mpi-fortran-wrappers.py (hunk 7 above)
Comment (dahongli) [reply mpi-proxy-split/mpi-wrappers/generate-mpi-fortran-wrappers.py:51]: Sure, I have added the comments.

**Review by gc00 — APPROVED** [review]
Excellent!  LGTM.
Thanks.

**Pull request merged**